src/jit/lower.cpp

```diff
@@ -2940,6 +2940,107 @@ GenTree* Lowering::LowerJTrue(GenTreeOp* jtrue)
     return nullptr;
 }
 
+//----------------------------------------------------------------------------------------------
+// LowerNodeCC: Lowers a node that produces a boolean value by setting the condition flags.
+//
+// Arguments:
+//     node - The node to lower
+//     condition - The condition code of the generated SETCC/JCC node
+//
+// Return Value:
+//     A SETCC/JCC node or nullptr if `node` is not used.
+//
+// Notes:
+//     This simply replaces `node`'s use with an appropiate SETCC/JCC node,
+//     `node` is not actually changed, except by having its GTF_SET_FLAGS set.
+//     It's the caller's responsibility to change `node` such that it only
+//     sets the condition flags, without producing a boolean value.
+//
+GenTreeCC* Lowering::LowerNodeCC(GenTree* node, GenCondition condition)
+{
+    // Skip over a chain of EQ/NE(x, 0) relops. This may be present either
```

Comment:
The code in `LowerSIMD` that this replaces also checked for `EQ/NE(x, 1)`. I added that code in #14027 but now I can't find any use for that extra complexity. Even if both the JIT and the C# compiler sometimes assume 0/1 bools when performing some optimizations, I've never seen a case that involves _testing_ bools and assuming 0/1, testing is always 0/non-zero.

Reply:
> I've never seen a case that involves testing bools and assuming 0/1, testing is always 0/non-zero.

Not sure what you would consider "testing", but for C# code there are a number of places where the generated IL doesn't strictly do `0` or `not-zero` checks. For example, C# returns `false` for `bool == bool` where the underlying bits are not identical and a switch statement can hit the default case for a bool which is not 0/1. I believe the csharplang repo had a few more oddities commented at one time.

Reply:
By testing I mean something like
```C#
if (ptest(x, y)) ...
if (!ptest(x, y)) ...
bool b = ptest(x, y)...
bool b = !ptest(x, y)...
```
I've never seen the C# compiler generating anything other than 0 compares in all these cases. In particular, logical negation (the subject of #21247) is implemented using `ceq(x, 0)` or `brfalse`.

Reply:
> I've never seen the C# compiler generating anything other than 0 compares in all these cases

Ah, I think the one case where it does an explicit `ceq(x, 1)` (rather than `ceq(x, 0`, `brfalse`, or `brtrue`) is for switch statements. I doubt such a code-pattern is that common, however.

In this case:
```csharp
switch (condition)
{
    case false:
        Console.WriteLine("0");
        break;

    case true:
        Console.WriteLine("1");
        break;

    default:
        Console.WriteLine("?");
        break;
}
```

becomes:
```ilasm
        IL_0000: ldarg.1
        IL_0001: brfalse.s IL_0009

        IL_0003: ldarg.1
        IL_0004: ldc.i4.1
        IL_0005: beq.s IL_0014

        IL_0007: br.s IL_001f

        IL_0009: ldstr "0"
        IL_000e: call void [mscorlib]System.Console::WriteLine(string)
        IL_0013: ret

        IL_0014: ldstr "1"
        IL_0019: call void [mscorlib]System.Console::WriteLine(string)
        IL_001e: ret

        IL_001f: ldstr "?"
        IL_0024: call void [mscorlib]System.Console::WriteLine(string)
        IL_0029: ret
```